src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionPools.cs

```diff
@@ -0,0 +1,347 @@
+                    // no other one holding the data connection.
+                    // in those error case, don't crash but return null. this method is TryCreate since caller expects it to return null
+                    // on such error situation.
+                    return null;
+                }
+
+                // get stream from service hub to communicate service specific information
+                // this is what consumer actually use to communicate information
+                var serviceStream = await Connections.RequestServiceAsync(_hubClient, serviceName, _hostGroup, _timeout, cancellationToken).ConfigureAwait(false);
+
+                return new JsonRpcConnection(_hubClient.Logger, callbackTarget, serviceStream, dataRpc);
+            }
+
+            private void Free(string serviceName, JsonRpcConnection connection)
+            {
+                using (_shutdownLock.DisposableRead())
+                {
+                    if (!_usePool)
+                    {
+                        // pool is not being used.
+                        connection.Dispose();
+                        return;
+                    }
+
+                    // queue must exist
+                    var queue = _pools[serviceName];
+                    if (queue.Count >= _maxPoolConnections)
+                    {
+                        // let the connection actually go away
+                        connection.Dispose();
+                        return;
+                    }
+
+                    // pool the connection
+                    queue.Enqueue(connection);
+                }
+            }
+
+            public void Shutdown()
+            {
+                using (_shutdownLock.DisposableWrite())
+                {
+                    // mark not to use pool
+                    _usePool = false;
+
+                    // let ref count this one is holding go
+                    _remotableDataRpc.Dispose();
+
+                    // let all connections in the pool to go away
+                    foreach (var kv in _pools)
+                    {
+                        while (kv.Value.TryDequeue(out var connection))
+                        {
+                            connection.Dispose();
+                        }
+                    }
+                }
+            }
+
+            private class PooledConnection : Connection
+            {
+                private readonly ConnectionPools _pools;
+                private readonly string _serviceName;
+                private readonly JsonRpcConnection _connection;
+
+                public PooledConnection(ConnectionPools pools, string serviceName, JsonRpcConnection connection)
+                {
+                    _pools = pools;
+                    _serviceName = serviceName;
+                    _connection = connection;
+                }
+
+                public override Task SetConnectionStateAsync(PinnedRemotableDataScope scope)
+                {
```

Comment:
nit.  consider => for all these forwarding methods.

Reply:
sure.